### ISSUE TITLE
Issue when destroying am::client

### ISSUE THREAD

**Issue by michieldwitte** [redboltz/async_mqtt#313]
We have a case where we need to close and reconnect to a broker and we stumbled on an issue.

I've been able to reduce it to the following code that gives the following error when built/run with `-fsanitize=address`  
(There's an MQTT broker running locally)

```
void test() {
    as::io_context ioc;
    auto exe = ioc.get_executor();
    // auto amcl = am::client<am::protocol_version::v3_1_1, am::protocol::mqtt>(exe); // Works
    as::co_spawn(
        exe,
        [&] () -> as::awaitable<void> {
            auto amcl = am::client<am::protocol_version::v3_1_1, am::protocol::mqtt>(exe); // heap-use-after-free

            auto [ec_und] = co_await am::async_underlying_handshake(
                amcl.next_layer(),
                "127.0.0.1",
                "1883",
                as::as_tuple(as::use_awaitable)
            );

            auto [ec_con, connack_opt] = co_await amcl.async_start(
                am::v3_1_1::connect_packet{
                    true,   // clean_session
                    0,      // keep_alive
                    ""
                },
                as::as_tuple(as::use_awaitable)
            );

            auto [ec_disconnect] = co_await amcl.async_disconnect(
                am::v3_1_1::disconnect_packet{},
                as::as_tuple(as::use_awaitable)
            );

            co_await amcl.async_close(
                as::as_tuple(as::use_awaitable)
            );

            co_return;
        },
        as::detached
    );
    ioc.run();
}
```

We get the following output of the AddressSanitizer
[address_sanitizer.log](https://github.com/user-attachments/files/15956772/address_sanitizer.log)

**Comment by redboltz** [comment]
After some analysis, I found the root issue.
endpoint (and basic_endpoint) use enable_shared_from_this approach. It is required to keep member variables' life time during async operation is running even if the user code endpoint object is scoped out.
client has an endpoint internally. So endpoint members life time is well managed by the endpoint itself.
However, the member variables of the client is not. I had looked over that case.
Your code example is this case.

I will fix it using enable_shared_from_this approach like endpoint.
It would cause the breaking change.

Your code example would become as follows:

```cpp
#include <async_mqtt/all.hpp>

namespace as = boost::asio;
namespace am = async_mqtt;

void test() {
    as::io_context ioc;
    auto exe = ioc.get_executor();
    // auto amcl = am::client<am::protocol_version::v3_1_1, am::protocol::mqtt>(exe); // Works
    as::co_spawn(
        exe,
        [&] () -> as::awaitable<void> {
            auto amcl = am::client<am::protocol_version::v3_1_1, am::protocol::mqtt>::create(exe); // heap-use-after-free

            auto [ec_und] = co_await am::async_underlying_handshake(
                amcl->next_layer(),
                "127.0.0.1",
                "1883",
                as::as_tuple(as::use_awaitable)
            );
            (void)ec_und;
            auto [ec_con, connack_opt] = co_await amcl->async_start(
                am::v3_1_1::connect_packet{
                    true,   // clean_session
                    0,      // keep_alive
                    ""
                },
                as::as_tuple(as::use_awaitable)
            );

            auto [ec_disconnect] = co_await amcl->async_disconnect(
                am::v3_1_1::disconnect_packet{},
                as::as_tuple(as::use_awaitable)
            );
            (void)ec_disconnect;

            co_await amcl->async_close(
                as::as_tuple(as::use_awaitable)
            );

            co_return;
        },
        as::detached
    );
    ioc.run();
}

int main() {
    am::setup_log(am::severity_level::trace);
    test();
}
```

I start fixing.

**Comment by redboltz** [comment]
Should be fixed by #315.
Please check it
